### PR TITLE
xds: add support for multiple xDS clients, for fallback

### PR DESCRIPTION
**Pull request by easwars**
This PR adds support for multiple xDS clients, as required by xDS fallback(A71).

Summary of changes:
- xDS client creation APIs now require a name to be specified
  - On the client side, the user's dial target is used, and on the server side a well-known dedicated key `#server` is used
  - Tests pass `t.Name` as the xDS client name
  - Also, added a `GetForTesting` method since it is no longer possible to simply call `New` and expect to get a reference to the singleton xDS client (as used by some tests to trigger resource not found errors)
  - Changes to the following resolvers to pass user's dial target during xDS client creation
    - `google_c2p`
    - `xds`
- xDS client changes to dumping resources
  - Removed the `DumpResources` method from the `XDSClient` interface
  - Added a `DumpResources` function to the `xdsclient` package that traverses all existing xDS clients, retrieves a dump from each of them, and adds the `client_scope` field to each one, and returns a response that contains resources from all the clients
- CSDS service implementation changes
  - No longer holds a reference to the xDS singleton. Instead invokes the `DumpResources` API provided by the `xdsclient` package (that retrieves configs from all xDS clients in the system).
  - E2E test now creates more than one xDS client and verifies that the `client_scope` field is set appropriately.
- Minor improvements to stringing functionality in the `bootstrap` package to improve test log readability.

Fixes https://github.com/grpc/grpc-go/issues/6899

#a71-xds-fallback

RELEASE NOTES: none

### CODE REVIEW

**Comment by easwars** [comment]
I probably need to add one more test where the xDS clients are talking to different management servers, and receiving different resources. But the review can start prior to that, and I should have that test written soonish.

**Comment by easwars** [comment]
`TestDumpResources` is failing because the test context is expiring before we test can read the expected config dump from the xDS clients. My guess at this point is that because the go-control-plane keeps resending resources when they are NACKed, the gRPC client is overwhelmed by ADS responses. But I have to confirm this theory. Will update when I have a breakthrough.

**Comment by easwars** [comment]
Tests are passing now. This is good to be looked at.

File: xds/internal/xdsclient/clientimpl_dump.go
Comment (zasweq) [inline xds/internal/xdsclient/clientimpl_dump.go:48]: Optional: I don't think this matters, but do we want this to be deterministic? This is related to the new Go heuristic? rule? Doug brought up for the nondeterministic metadata parsing: "Maps are printed in a consistent order, sorted by the values of the keys." I know this isn't really printing it, just building out a response.

File: xds/server.go
Comment (zasweq) [inline xds/server.go:47]: xdsClientServerKey maybe?

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted.go:43]: Optional: this function doesn't necessarily close the clientImpl for the name. It decrements the ref and if it's the last ref, it closes/removes it from global map. Should this be called clientRefCountedDecrementRef instead? I see it was already called this.

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted.go:90]: I'm having a hard time understanding why it's always at index 0 with respect to this dimension here: "Current gRPC xDS implementations only use the first xDS configuration server listed in the bootstrap JSON document. Fallback implementation requires changes to use all servers listed in order from highest to lowest priority. Priority is assigned based on server order in bootstrap.json." Is this handled at a higher layer, and once called into this it is expected to only have one xDS Server in the config?

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted_test.go:56]: Fix what? Here and below.

File: xds/internal/xdsclient/clientimpl_dump.go
Comment (zasweq) [inline xds/internal/xdsclient/clientimpl_dump.go:42]: Nit, drop the end of this, already implied from the return value.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted_test.go:118]: Now that this is no longer a singleton, should we have some tests for the new dimension of numerous clients per name and ref counting/closing across clients of that dimension?

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted_test.go:39]: Here and below: this comment needs updating, no longer singleton.

File: xds/csds/csds_e2e_test.go
Comment (zasweq) [inline xds/csds/csds_e2e_test.go:141]: Top level comment please with stuff like two xDS Clients etc.

File: xds/csds/csds_e2e_test.go
Comment (zasweq) [inline xds/csds/csds_e2e_test.go:358]: This adds testing scope for the numerous xDS Clients and client scope. However, they point to the same resources. Should we test the clients pointing to different resources to make sure all that works correctly? Also another dimension alongside different resources is some resources in one client being in an error state while others not, alongside just a raw different state, which I also think is worth testing.

File: xds/csds/csds_e2e_test.go
Comment (zasweq) [inline xds/csds/csds_e2e_test.go:376]: Optional: top level comments that this polls and what it asserts? And stuff like sorting client scope and resources for determinism?

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted.go:61]: Top level comment of get or create if not present? Or change name?

File: xds/internal/xdsclient/client_new.go
Comment (zasweq) [inline xds/internal/xdsclient/client_new.go:139]: Why did you chose to split this operation up for testing instead of like newRefCountedWithConfig does a get or create in one operation? Is it for ease of use where you create with knobs and then can access by name?

**Review by zasweq — COMMENTED** [review]
Implementation looks good overall, mostly just style/naming nits. Some questions about scaling up tests.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted_test.go:118]: Added a test. Thank you.

File: xds/csds/csds_e2e_test.go
Comment (easwars) [reply xds/csds/csds_e2e_test.go:141]: Done.

File: xds/csds/csds_e2e_test.go
Comment (easwars) [reply xds/csds/csds_e2e_test.go:358]: I have tests in the `xdsclient` package which creates xDS clients talking to different servers. Adding the same tests here, I feel does not add much value, since this already tests the CSDS functionality thoroughly.

File: xds/csds/csds_e2e_test.go
Comment (easwars) [reply xds/csds/csds_e2e_test.go:376]: Done.

**Comment by easwars** [comment]
Looks like there is a flake in the csds e2e test. I will continue to look at that, but this is still good for a second pass. Thanks.

File: xds/internal/xdsclient/client_new.go
Comment (zasweq) [reply xds/internal/xdsclient/client_new.go]: Ah right this is arbitrary in tests used for map key, this is in reference to production usage. Deleting it fully sgtm.

File: xds/internal/xdsclient/clientimpl_dump.go
Comment (zasweq) [reply xds/internal/xdsclient/clientimpl_dump.go:47]: Mark mentioned in chat that CSDS has no guarantee of ordering, and the deterministic map thing is Go specific, so this is ok.

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [reply xds/internal/xdsclient/client_refcounted.go:43]: I don't feel strongly. Is it because semantically a ref counted thing is known to do this behind the scenes?

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [reply xds/internal/xdsclient/client_refcounted.go:93]: Right, I meant what you logged. My point was that once you actually implement fallback, does reading at index 0 still make sense for logging? Or will you change this once that change happens, or is that irrelevant because the caller of this handles fallback?

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [reply xds/internal/xdsclient/client_refcounted.go:64]: Same question as for the Close, is the reason to keep these names this happens implicitly with objects that are ref counted?

File: xds/internal/xdsclient/client_new.go
Comment (zasweq) [reply xds/internal/xdsclient/client_new.go:140]: Alright that's fair I like that.

File: xds/server.go
Comment (zasweq) [reply xds/server.go]: Nice I like the consts name.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [reply xds/internal/xdsclient/client_refcounted_test.go]: Oh you're intention was to make it the test name to uniquely identify I see.

File: xds/csds/csds_e2e_test.go
Comment (zasweq) [reply xds/csds/csds_e2e_test.go:376]: Hmmm fair point, this tests the logic for errors/different CSDS operaitons.

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted.go:63]: Nit: "gets a reference to the existing xDS Client, or creates one if one does not exist already" or something alongside those lines for first context. To explicitly call out what happens if the xDS Client already exists for this name.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted_test.go:124]: Nit: switch "are released, the client is closed" to something that emphasizes the plurality of the clients maybe "are released for a client, only that specific client is closed" or something along those lines.

**Review by zasweq — APPROVED** [review]
Looks great!

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (zasweq) [inline xds/internal/xdsclient/client_refcounted_test.go:218]: Optional: here and the close for 2: the test above tests the idempotency implemented through the once func. Maybe call some of these twice?

File: xds/csds/csds_e2e_test.go
Comment (zasweq) [reply xds/csds/csds_e2e_test.go:376]: Just make sure maybe manually in the future CSDS looks right with multiple clients that point to different servers that have different resource states/errors for those resources.

File: xds/internal/xdsclient/client_refcounted.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted.go:43]: Yes, this would have simply been `close` or `clientClose`. But because we have different client-ish objects here, the `XDSClient` interface, the `clientImpl` which is the implementation, and this one, which is the ref counted wrapper, the name ended up containing the `RefCounted` part.

On second thoughts though, we might be able to get rid of this type completely and move the reference counting functionality to either the `clientImpl` type or we could come up with a generic reference counted type that can wrap anything. I will file an issue on github to explore this further (don't want to do it as part of this PR, which is already huge).

File: xds/internal/xdsclient/client_refcounted.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted.go:63]: Done.

File: xds/internal/xdsclient/client_refcounted.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted.go:64]: Explained the reasoning in the comment for close.

File: xds/internal/xdsclient/client_refcounted.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted.go:93]: I don't think it really matters that we include all the xds server names in this log message. But I changed the log message to indicate that the server being printed is the primary.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted_test.go:124]: Done.

File: xds/internal/xdsclient/client_refcounted_test.go
Comment (easwars) [reply xds/internal/xdsclient/client_refcounted_test.go:218]: Sure. Done.

File: xds/internal/xdsclient/client_refcounted.go
Comment (zasweq) [reply xds/internal/xdsclient/client_refcounted.go:43]: The latter sounds good. I remember from C++ they have a ptr type that essentially does ref counting/freeing for you so it would be a type in Go that serves similar functionality.

**Comment by easwars** [comment]
The CSDS test is failing on Github Actions because of the way go-control-plane continuously resends NACKed resources. This results in a busy loop on the xdsClient where it receives a resource and NACKs it repeatedly. This results in starvation of the `DumpResources` call from the CSDS service.

I'm planning to put this PR on hold for now, fix https://github.com/grpc/grpc/issues/34099 and get back here.

**Comment by easwars** [comment]
@zasweq : Based on our offline discussion, I made a commit that skips the test on arm64. Waiting for GA to pass before I merge this. That way, I can work on the other PR implementing ADS stream flow control, and once I have that, I will be able to tweak this test and get it to work.

**Comment by zasweq** [comment]
Sgtm, thinking about it I agree with Doug. To minimize the chance of merge conflicts probably best to get this in sooner rather than later.

**Pull request merged**
